### PR TITLE
SOLR-15258: Implement FileSystemConfigSetService and add TestConfigSe…

### PR DESCRIPTION
**Pull request by NazerkeBS**
…tService

https://issues.apache.org/jira/browse/SOLR-XXXXX

# Description

Please provide a short description of the changes you're making with this pull request.

# Solution

Please provide a short description of the approach taken to implement your solution.

# Tests

Please describe the tests you've developed or run to confirm this patch implements the feature or solves the problem.

# Checklist

Please review the following and check all that apply:

- [ ] I have reviewed the guidelines for [How to Contribute](https://wiki.apache.org/solr/HowToContribute) and my code conforms to the standards described there to the best of my ability.
- [ ] I have created a Jira issue and added the issue ID to my pull request title.
- [ ] I have given Solr maintainers [access](https://help.github.com/en/articles/allowing-changes-to-a-pull-request-branch-created-from-a-fork) to contribute to my PR branch. (optional but recommended)
- [ ] I have developed this patch against the `main` branch.
- [ ] I have run `./gradlew check`.
- [ ] I have added tests for my changes.
- [ ] I have added documentation for the [Reference Guide](https://github.com/apache/solr/tree/main/solr/solr-ref-guide)

### CODE REVIEW

**Comment by NazerkeBS** [comment]
@dsmiley, maybe I should update ZkConfigSetService#checkConfigExists in this PR  to check if the config dir is not empty and contains solrconfig.xml file ?

**Comment by dsmiley** [comment]
> maybe I should update ZkConfigSetService#checkConfigExists in this PR to check if the config dir is not empty and contains solrconfig.xml file ?

No; it is unrelated to this PR.

**Review by dsmiley — CHANGES_REQUESTED** [review]
My main concern is that the HTTP API for ConfigSets could be abused by an attacker to edit other configSets if the configName contained "../otherconfig" type stuff.  If the getConfigDir method in here were used universally by the other methods and if it validated that configName is indeed right below configSetBase then I think this concern would be addressed.

**Comment by janhoy** [comment]
Jira SOLR-15258 is closed, can this PR also be closed?

**Comment by dsmiley** [comment]
SOLR-15258 is a refactoring that we might say can include this or maybe not; it doesn't matter I suppose.  It's 9.0 only so there is more fluidity.  I suppose a separate issue feels better.

**Comment by janhoy** [comment]
The answer made me no wiser :) To me it looked like this is committed elsewhere without being closed. But you're saying this PR should get another JIRA number and remains to be merged? I did not look closely.

**Comment by NazerkeBS** [comment]
@janhoy this PR is not committed before. I think jira number is confusing  here as SOLR-15258 is already closed. 
I will create a jira issue and link this PR with that. But still it remains to be merged.

**Comment by janhoy** [comment]
Thanks for the clarification :)

File: solr/core/src/java/org/apache/solr/cloud/ZkConfigSetService.java [hunk 1]
```diff
@@ -232,15 +230,20 @@ public void setConfigMetadata(String configName, Map<String, Object> data) throw
...
+    byte[] data = null;
     try {
-      @SuppressWarnings("unchecked")
-      Map<String, Object> data =
-          (Map<String, Object>)
-              Utils.fromJSON(zkClient.getData(CONFIGS_ZKNODE + "/" + configName, null, null, true));
-      return data;
+      data = zkClient.getData(CONFIGS_ZKNODE + "/" + configName, null, null, true);
+    } catch (KeeperException.NoNodeException e) {
+      return new HashMap<>();
```
Comment (epugh) [inline solr/core/src/java/org/apache/solr/cloud/ZkConfigSetService.java:237]: I like we simplified the lines.   Should we return from the exception, or just catch it since below we check `data == null` and then return?

File: solr/core/src/java/org/apache/solr/cloud/ZkConfigSetService.java [hunk 2]
```diff
@@ -232,15 +230,20 @@ public void setConfigMetadata(String configName, Map<String, Object> data) throw
...
     } catch (KeeperException | InterruptedException e) {
       throw new IOException("Error getting config metadata", SolrZkClient.checkInterrupted(e));
     }
+    if (data == null) {
+      return new HashMap<>();
+    }
+    @SuppressWarnings("unchecked")
```
Comment (epugh) [inline solr/core/src/java/org/apache/solr/cloud/ZkConfigSetService.java:244]: likewise, should this be an if/else ?   I don't always know if stylisitically if we are okay with mulitple returns from a method?

File: solr/core/src/java/org/apache/solr/cloud/ZkConfigSetService.java (hunk 1 above)
Comment (NazerkeBS) [reply solr/core/src/java/org/apache/solr/cloud/ZkConfigSetService.java:237]: previously NoNodeException is not checked.

**Review by dsmiley — COMMENTED** [review]
@NazerkeBS I see you added stuff outside of FileSystemConfigSetService.  Can you summarize those changes please?  What is the intent behind them?  I suspect they are unrelated to FileSystemConfigSetService and thus should be a separate PR (probably no JIRA).

**Comment by NazerkeBS** [comment]
> @NazerkeBS I see you added stuff outside of FileSystemConfigSetService. Can you summarize those changes please? What is the intent behind them? I suspect they are unrelated to FileSystemConfigSetService and thus should be a separate PR (probably no JIRA).

@dsmiley please see the comments above

File: solr/core/src/java/org/apache/solr/core/FileSystemConfigSetService.java [hunk 3]
```diff
@@ -60,74 +79,192 @@ public String configSetName(CoreDescriptor cd) {
...
+    }
   }
 
   @Override
   public void setConfigMetadata(String configName, Map<String, Object> data) throws IOException {
-    throw new UnsupportedOperationException();
+    // store metadata in .metadata.json file
+    Path metadataPath = getConfigDir(configName).resolve(METADATA_FILE);
+    Files.write(metadataPath, Utils.toJSON(data));
+    Files.isHidden(metadataPath);
```
Comment (dsmiley) [inline solr/core/src/java/org/apache/solr/core/FileSystemConfigSetService.java:159]: This line does nothing

File: solr/core/src/java/org/apache/solr/core/FileSystemConfigSetService.java [hunk 4]
```diff
@@ -60,74 +79,192 @@ public String configSetName(CoreDescriptor cd) {
...
+            if (!Files.isHidden(file)) {
+              filePaths.add(configDir.relativize(file).toString());
+              return FileVisitResult.CONTINUE;
+            }
+            return FileVisitResult.CONTINUE;
+          }
+
+          @Override
+          public FileVisitResult postVisitDirectory(Path dir, IOException ioException) {
+            if (!dir.getFileName().toString().equals(configName)) {
```
Comment (dsmiley) [inline solr/core/src/java/org/apache/solr/core/FileSystemConfigSetService.java:254]: if the intermediate directory had a name that coincidentally was the same as that of the configSet, this would not do what we want.  I think you could skip this check and instead check that the relatived-dir string path is not empty?

File: solr/core/src/java/org/apache/solr/cloud/ZkConfigSetService.java (hunk 2 above)
Comment (dsmiley) [reply solr/core/src/java/org/apache/solr/cloud/ZkConfigSetService.java:247]: We have no coding standard apart from the [Google Java Style](https://google.github.io/styleguide/javaguide.html) which makes no mentions of multiple returns.  FWIW, I like multiple returns!

File: solr/core/src/java/org/apache/solr/core/ConfigSetService.java [hunk 5]
```diff
@@ -346,7 +342,7 @@ protected IndexSchema createIndexSchema(
    * Returns a modification version for the schema file. Null may be returned if not known, and if
    * so it defeats schema caching.
    */
-  protected abstract Long getCurrentSchemaModificationVersion(
+  public abstract Long getCurrentSchemaModificationVersion(
```
Comment (dsmiley) [inline solr/core/src/java/org/apache/solr/core/ConfigSetService.java:345]: This was made public so that a custom ConfigSetService can delegate this method call.

File: solr/core/src/java/org/apache/solr/core/FileSystemConfigSetService.java [hunk 6]
```diff
@@ -60,74 +79,192 @@ public String configSetName(CoreDescriptor cd) {
...
+    Path metadataPath = getConfigDir(configName).resolve(METADATA_FILE);
+    byte[] data = null;
+    try {
+      data = Files.readAllBytes(metadataPath);
+    } catch (NoSuchFileException e) {
+      return new HashMap<>();
+    }
+    if (data == null) {
+      return new HashMap<>();
+    }
```
Comment (dsmiley) [inline solr/core/src/java/org/apache/solr/core/FileSystemConfigSetService.java:174]: Utils.fromJSON already checks for null param and returns an empty map

File: solr/core/src/java/org/apache/solr/core/FileSystemConfigSetService.java [hunk 7]
```diff
@@ -60,74 +79,192 @@ public String configSetName(CoreDescriptor cd) {
...
   @Override
   public Map<String, Object> getConfigMetadata(String configName) throws IOException {
-    throw new UnsupportedOperationException();
+    // get metadata from .metadata.json file
+    Path metadataPath = getConfigDir(configName).resolve(METADATA_FILE);
+    byte[] data = null;
+    try {
+      data = Files.readAllBytes(metadataPath);
+    } catch (NoSuchFileException e) {
+      return new HashMap<>();
```
Comment (dsmiley) [inline solr/core/src/java/org/apache/solr/core/FileSystemConfigSetService.java:170]: `Collections.emptyMap()` -- no need for a new mutable map

File: solr/core/src/test/org/apache/solr/core/TestConfigSetService.java [hunk 8]
```diff
@@ -0,0 +1,126 @@
...
+    configSetService.setConfigMetadata(configName, Collections.singletonMap("trusted", true));
+    metadata = configSetService.getConfigMetadata(configName);
+    assertTrue(metadata.containsKey("trusted"));
+
+    configSetService.setConfigMetadata(configName, Collections.singletonMap("foo", true));
+    assertFalse(configSetService.getConfigMetadata(configName).containsKey("trusted"));
+    assertTrue(configSetService.getConfigMetadata(configName).containsKey("foo"));
+
+    List<String> configFiles = configSetService.getAllConfigFiles(configName);
+    assertTrue(configFiles.size() == 5);
```
Comment (dsmiley) [inline solr/core/src/test/org/apache/solr/core/TestConfigSetService.java:112]: can you assert the list contents so we can see the expected strings?  I ask because this method is somewhat quirky / specific about the response, e.g. sorted & has intermediate dirs

**Review by dsmiley — COMMENTED** [review]
Do the FileSystemConfigSetService changes allow a standalone Solr node to respond to ConfigSet API requests that modify configSets?

File: solr/core/src/test/org/apache/solr/core/TestFileSystemConfigSetService.java [hunk 9]
```diff
@@ -0,0 +1,93 @@
...
+  }
+
+  @AfterClass
+  public static void afterClass() throws Exception {
+    PathUtils.deleteDirectory(configSetBase);
+    fileSystemConfigSetService = null;
+  }
+
+  @Test
+  public void testUploadConfig() throws IOException {
```
[the remark this answers is not in the document]
Comment (NazerkeBS) [reply solr/core/src/test/org/apache/solr/core/TestFileSystemConfigSetService.java:52]: I checked with "/" input as a configName and  there is access denied FilePermission exception.

[the remark this answers is not in the document]
Comment (dsmiley) [reply solr/core/src/test/org/apache/solr/core/TestFileSystemConfigSetService.java:52]: That sounds like you were stopped by the Java security manager or OS.  If you change the path to someplace the test can write to then I think it will show the flaw

[the remark this answers is not in the document]
Comment (NazerkeBS) [reply solr/core/src/test/org/apache/solr/core/TestFileSystemConfigSetService.java:52]: Yes, it is java security manager. I checked relative configName path to configSetBase and throw IOException.

[the remark this answers is not in the document]
Comment (NazerkeBS) [reply solr/core/src/test/org/apache/solr/core/TestFileSystemConfigSetService.java:52]: When it comes to FileSystemConfigSetService.locateInstanceDir, in case the presence of configset, we still need the dir for the configset.

File: solr/core/src/java/org/apache/solr/core/FileSystemConfigSetService.java [hunk 10]
```diff
@@ -148,4 +284,13 @@ protected Long getCurrentSchemaModificationVersion(
       return null; // debatable; we'll see an error soon if there's a real problem
     }
   }
+
+  protected Path getConfigDir(String configName) throws IOException {
+    String configSetDir = configSetBase.toFile().getCanonicalPath() + File.separator;
+    File file = new File(configSetBase.toString(), configName);
```
[the remark this answers is not in the document]
Comment (dsmiley) [reply solr/core/src/java/org/apache/solr/core/FileSystemConfigSetService.java:290]: Please avoid File; Path is Java's newer & better replacement.   It has methods to do what you want to do.

**Comment by dsmiley** [comment]
Tip:  Follow the [Google Java Format configuration settings for your IDE](https://github.com/google/google-java-format).  Furthermore, I suggest telling IntelliJ to automatically apply formatting on commit.  This will reduce the amount of "tidy"-ing you do a lot!

File: solr/core/src/java/org/apache/solr/core/FileSystemConfigSetService.java [hunk 11]
```diff
@@ -287,9 +287,11 @@ public Long getCurrentSchemaModificationVersion(
...
 
   protected Path getConfigDir(String configName) throws IOException {
-    String configSetDirPath = configSetBase.toFile().getCanonicalPath() + File.separator;
-    String path = Paths.get(configSetBase.toString(), configName).toFile().getCanonicalPath();
-    if (!path.startsWith(configSetDirPath)) {
+    String path =
+        Paths.get(configSetBase.toString(), FilenameUtils.getName(configName))
+            .toFile()
+            .getCanonicalPath();
+    if (!path.endsWith(configName)) {
```
Comment (dsmiley) [inline solr/core/src/java/org/apache/solr/core/FileSystemConfigSetService.java:294]: Don't we want to check that the path *startsWith* the configSetBase?  I don't think we should care what it ends with.  Otherwise we could read a path like "../../../somewhereElse/configName" that the caller shouldn't have access to.

File: solr/core/src/java/org/apache/solr/core/FileSystemConfigSetService.java [hunk 12]
```diff
@@ -287,9 +287,11 @@ public Long getCurrentSchemaModificationVersion(
   }
 
   protected Path getConfigDir(String configName) throws IOException {
-    String configSetDirPath = configSetBase.toFile().getCanonicalPath() + File.separator;
-    String path = Paths.get(configSetBase.toString(), configName).toFile().getCanonicalPath();
-    if (!path.startsWith(configSetDirPath)) {
+    String path =
+        Paths.get(configSetBase.toString(), FilenameUtils.getName(configName))
+            .toFile()
+            .getCanonicalPath();
```
Comment (dsmiley) [inline solr/core/src/java/org/apache/solr/core/FileSystemConfigSetService.java:293]: Why getCanonicalPath?

File: solr/core/src/java/org/apache/solr/core/FileSystemConfigSetService.java [hunk 13]
```diff
@@ -287,9 +287,11 @@ public Long getCurrentSchemaModificationVersion(
   }
 
   protected Path getConfigDir(String configName) throws IOException {
-    String configSetDirPath = configSetBase.toFile().getCanonicalPath() + File.separator;
-    String path = Paths.get(configSetBase.toString(), configName).toFile().getCanonicalPath();
-    if (!path.startsWith(configSetDirPath)) {
+    String path =
+        Paths.get(configSetBase.toString(), FilenameUtils.getName(configName))
```
Comment (dsmiley) [inline solr/core/src/java/org/apache/solr/core/FileSystemConfigSetService.java:291]: If configName were "foo/bar", there shouldn't be anything fundamentally wrong with that.

File: solr/core/src/java/org/apache/solr/core/FileSystemConfigSetService.java (hunk 12 above)
Comment (NazerkeBS) [reply solr/core/src/java/org/apache/solr/core/FileSystemConfigSetService.java]: to remove intermediate `.` or `..` from absolute Path.

File: solr/core/src/java/org/apache/solr/core/FileSystemConfigSetService.java [hunk 14]
```diff
@@ -287,11 +285,9 @@ public Long getCurrentSchemaModificationVersion(
...
 
   protected Path getConfigDir(String configName) throws IOException {
-    String path =
-        Paths.get(configSetBase.toString(), FilenameUtils.getName(configName))
-            .toFile()
-            .getCanonicalPath();
-    if (!path.endsWith(configName)) {
+    String configSetBasePath = configSetBase.toString();
+    String path = Path.of(configSetBasePath, configName).toFile().getCanonicalPath();
+    if (!path.startsWith(configSetBasePath)) {
```
Comment (dsmiley) [inline solr/core/src/java/org/apache/solr/core/FileSystemConfigSetService.java:290]: No, I meant Path.startsWith(path).  In general, hold paths as Path not String nor File.  It's typed, has lots of useful methods, and easier to "get right" whereas String processing in particular tends to have more subtle bugs.

File: solr/core/src/java/org/apache/solr/core/FileSystemConfigSetService.java [hunk 15]
```diff
@@ -148,4 +283,12 @@ protected Long getCurrentSchemaModificationVersion(
       return null; // debatable; we'll see an error soon if there's a real problem
     }
   }
+
+  protected Path getConfigDir(String configName) throws IOException {
+    Path path = configSetBase.resolve(configName).normalize();
```
Comment (dsmiley) [inline solr/core/src/java/org/apache/solr/core/FileSystemConfigSetService.java:288]:
```suggestion
    // startsWith works simply; we must normalize()
    Path path = configSetBase.resolve(configName).normalize();
```

**Review by dsmiley — COMMENTED** [review]
So I was wrong on thinking normalize() wasn't necessary.  startsWith is simple; won't interpret non-normalized stuff smartly or anything.
